packages/notebook/src/executionindicator.tsx

```diff
@@ -0,0 +1,490 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import { ISessionContext } from '@jupyterlab/apputils';
+
+import { ITranslator, nullTranslator } from '@jupyterlab/translation';
+import React from 'react';
+import { interactiveItem, ProgressCircle } from '@jupyterlab/statusbar';
+
+import {
+  circleIcon,
+  LabIcon,
+  offlineBoltIcon,
+  VDomModel,
+  VDomRenderer
+} from '@jupyterlab/ui-components';
+
+import { Notebook } from './widget';
+import { KernelMessage } from '@jupyterlab/services';
+
+/**
+ * A react functional component for rendering execution indicator.
+ */
+export function ExecutionIndicatorComponent(
+  props: ExecutionIndicatorComponent.IProps
+): React.ReactElement<ExecutionIndicatorComponent.IProps> {
+  const translator = props.translator || nullTranslator;
+  const trans = translator.load('jupyterlab');
+  const state = props.state;
+  const showOnToolBar = props.displayOption.showOnToolBar;
+  const showProgress = props.displayOption.showProgress;
+  const tooltipClass = showOnToolBar ? 'down' : 'up';
+  const emptyDiv = <div></div>;
+
+  if (!state) {
+    return emptyDiv;
+  }
+
+  const kernelStatus = state.kernelStatus;
+  const circleIconProps: LabIcon.IProps = {
+    alignSelf: 'normal',
+    height: '24px'
+  };
+  const time = state.totalTime;
+
+  const scheduledCellNumber = state.scheduledCellNumber || 0;
+  const remainingCellNumber = state.scheduledCell.size || 0;
+  const executedCellNumber = scheduledCellNumber - remainingCellNumber;
+  let percentage = (100 * executedCellNumber) / scheduledCellNumber;
+  let displayClass = showProgress ? '' : 'hidden';
+  if (!showProgress && percentage < 100) {
+    percentage = 0;
+  }
+
+  const progressBar = (percentage: number) => (
+    <ProgressCircle progress={percentage} width={16} height={24} />
+  );
+  const titleFactory = (prefix: string, title: string) =>
+    trans.__(`${prefix} ${title[0].toUpperCase() + title.slice(1)}`);
```

Comment:
Thank you for very useful comments! The statuses used with this function are taken from `ISessionContext.kernelDisplayStatus`, and the prefixes are "Kernel" or "Kernel status", will they work with translation bundle?

Reply:
No, the translation won't work here if they are combined together as above. Three things need to happen:
- you need to pass already translated prefix, e.g. `titleFactory(trans.__('Kernel'), status);`
- re-use the map introduced in #10738 like `translatedStatus = statusNames[status];` you will need to move it somewhere else from the current location to export it and you could also fix #10807 by the way (in order to add the missing translations to already released 3.1 I added two copies of this Record mapping to avoid any API breakage, but since you are targeting 4.0 you are free to move stuff around).
- change the concatenation from ES6 template string to gettext: `trans.__('%1 %2', prefix, translatedStatus);` - think right-to-left languages - those will want to flip the order!
 
It would be probably better to hard-code prefixes to give translators more context to work with, like: `trans.__('Kernel %2', translatedStatus);` and `trans.__('Kernel status %2', translatedStatus);` but this is not a strong opinion.

Reply:
> you need to pass already translated prefix

(or use a Record mapping like `statusNames`)

Reply:
can you take a look at https://github.com/jupyterlab/jupyterlab/pull/10730/commits/c4a5d7d5c16d7c26d02897011bcf8eb78cb6a3d9 for the translation fixes?